### PR TITLE
use new double_pop_list to replace neighborset; 

### PR DESCRIPTION
**Pull request by alwayslove2013**
issue: #343 

- Use double_pop_list(`NeighborSetDoublePopList`) to replace single_pop_list(`NeighborSet`); 
- Update alpha-candidates strategy, alpha = filte_ratio / 2; 
- BitsetView's count only calculated once.

Results:
Tested directly via pyknowhere, cohere 1M * 768dim, cosine, single segment.
- recall remains stable with different filter_ratio.
- low filtering - qps improved. qps gap between filter 0 and 0.01 is smoothed out. 
- high filtering - qps is sacrificed to keep recall stable.
image

### CODE REVIEW

File: src/common/index.cc [hunk 1]
```diff
@@ -68,13 +68,14 @@ Index<T>::Add(const DataSet& dataset, const Json& json) {
 
 template <typename T>
 inline expected<DataSetPtr>
-Index<T>::Search(const DataSet& dataset, const Json& json, const BitsetView& bitset) const {
+Index<T>::Search(const DataSet& dataset, const Json& json, const BitsetView& _bitset) const {
```
Comment (PwzXxm) [inline src/common/index.cc:71]: Try not to define variables beginning with underscores. These may be reserved by OS and others.

File: thirdparty/hnswlib/hnswlib/neighbor_double_pop_list.h [hunk 2]
```diff
@@ -0,0 +1,204 @@
+
```
Comment (PwzXxm) [inline thirdparty/hnswlib/hnswlib/neighbor_double_pop_list.h:1]: extra space

File: thirdparty/hnswlib/hnswlib/neighbor_double_pop_list.h [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
```
Comment (PwzXxm) [inline thirdparty/hnswlib/hnswlib/neighbor_double_pop_list.h:2]: Do we still need `NeighborSet` in `neighbor.h`? Is it used anywhere else? If not, maybe clean it up?

File: thirdparty/hnswlib/hnswlib/hnswalg.h [hunk 4]
```diff
@@ -329,6 +329,7 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
             metric_hops++;
             metric_distance_computations += size;
         }
+        float kAlpha = bitset.filter_ratio() / 2.0f;
```
Comment (PwzXxm) [inline thirdparty/hnswlib/hnswlib/hnswalg.h:332]: Comment on why this change plz

File: src/common/index.cc (hunk 1 above)
Comment (alwayslove2013) [reply src/common/index.cc]: fixed.

File: thirdparty/hnswlib/hnswlib/neighbor_double_pop_list.h (hunk 2 above)
Comment (alwayslove2013) [reply thirdparty/hnswlib/hnswlib/neighbor_double_pop_list.h]: fixed.

File: thirdparty/hnswlib/hnswlib/neighbor_double_pop_list.h (hunk 3 above)
Comment (alwayslove2013) [reply thirdparty/hnswlib/hnswlib/neighbor_double_pop_list.h]: removed `NeighborSet`.
moved `NeighborSetDoublePopList` to `neighbor.h`.

File: thirdparty/hnswlib/hnswlib/neighbor.h [hunk 5]
```diff
@@ -28,83 +28,116 @@ struct Neighbor {
...
-        if (size_ == capacity_ && disqualified) {
-            disqualified->push(data_[size_ - 1]);
+        if (data_[lo].distance >= distance) {
+            return lo;
+        }
+        if (data_[hi - 1].distance <= distance) {
+            return hi;
         }
-        int lo = 0, hi = size_;
         while (lo < hi) {
```
Comment (alexanderguzhva) [inline thirdparty/hnswlib/hnswlib/neighbor.h:46]: can this loop be replaced with either `std::upper_bound()` or `std::lower_bound()`?

Comment (alwayslove2013) [reply thirdparty/hnswlib/hnswlib/neighbor.h]: Good idea~ I will use `upper_bound`.
